crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -124,25 +149,59 @@ async fn app_validation_workflow_inner(
+                            if let Dependency::Null = dependency {
+                                put_integrated(&mut txn, op_hash, ValidationStatus::Valid)?;
+                            } else {
+                                put_integration_limbo(&mut txn, op_hash, ValidationStatus::Valid)?;
+                            }
                         }
                         Outcome::AwaitingDeps(deps) => {
+                            awaiting += 1;
                             let status = ValidationLimboStatus::AwaitingAppDeps(deps);
                             put_validation_limbo(&mut txn, op_hash, status)?;
                         }
                         Outcome::Rejected(_) => {
-                                tracing::warn!("Received invalid op! Warrants aren't implemented yet, so we can't do anything about this right now, but be warned that somebody on the network has maliciously hacked their node.\nOp: {:?}", op_light);
-                            put_integration_limbo(&mut txn, op_hash, ValidationStatus::Rejected)?;
+                            rejected += 1;
+                            tracing::warn!("Received invalid op! Warrants aren't implemented yet, so we can't do anything about this right now, but be warned that somebody on the network has maliciously hacked their node.\nOp: {:?}", op_light);
+                            if let Dependency::Null = dependency {
+                                put_integrated(&mut txn, op_hash, ValidationStatus::Rejected)?;
+                            } else {
+                                put_integration_limbo(&mut txn, op_hash, ValidationStatus::Rejected)?;
+                            }
                         }
                     }
                 }
-                WorkflowResult::Ok(total)
+                WorkflowResult::Ok((total, awaiting, rejected))
             })
             .await?;
         total += t;
+        if let (Some(start), Some(round_time)) = (start, &mut round_time) {
+            let round_el = round_time.elapsed();
+            *round_time = std::time::Instant::now();
+            let avg_ops_ps = total as f64 / start.elapsed().as_micros() as f64 * 1_000_000.0;
+            let ops_ps = t as f64 / round_el.as_micros() as f64 * 1_000_000.0;
+            tracing::info!(
+                "App validation is saturated. Util {:.2}%. OPS/s avg {:.2}, this round {:.2}",
+                (start_len - total) as f64 / NUM_CONCURRENT_OPS as f64 * 100.0,
+                avg_ops_ps,
+                ops_ps
+            );
+        }
+        tracing::debug!(
+            "{} committed, {} awaiting sys dep, {} rejected. {} committed this round",
+            t,
+            a,
+            r,
+            total
+        );
     }
+    jh.await?;
     tracing::debug!("accepted {} ops", total);
-    Ok(WorkComplete::Complete)
+    Ok(if saturated {
+        WorkComplete::Incomplete
+    } else {
+        WorkComplete::Complete
+    })
```

Comment:
Just noting that this function has nearly doubled in size, from 60 lines to 120. Not sure if there was a clear place to break it up, but could be good to think about in the future.